### PR TITLE
[Perf] Pre-allocate pinned CPU buffers for sequential offloading

### PR DESCRIPTION
**Pull request by ZhanqiuHu**
**Ref #1686**: 
Model-level CPU offloading was slower than diffusers due to repeated per-parameter pin_memory() calls during GPU→CPU transfers.

## TL;DR

**Code is ready. Appreciate any feedback. Thanks!**

Add `PinnedWeightsStore` that pre-allocates page-locked CPU tensors at enable() time, so offloading copies directly into pinned memory (single DMA) instead of module.to("cpu") followed by per-parameter pin_memory(). Also pre-pin DiT parameters on CPU at initialization for faster CPU-to-GPU transfers.

| Config | Before | After | diffusers |
|--------|--------|-------|-----------|
| TP=1 + cpu-offload | 30.65s | **24.85s (19% faster)** | 25.89s |

## Test Plan

1. **Existing CI test**: `tests/e2e/offline_inference/test_diffusion_cpu_offload.py`
   — verifies memory savings and no crash with `riverclouds/qwen_image_random`.

```bash
python -m pytest tests/e2e/offline_inference/test_diffusion_cpu_offload.py -v
```

2. **Bitwise correctness** (FLUX.1-dev, manual — model is too large for CI):
   Generate with identical config (seed=42, 50 steps, 1024×1024) with and
   without `--enable-cpu-offload`, then compare output pixels. 

<details>
<summary>Reproduction commands</summary>

```bash
# with offload
python examples/offline_inference/text_to_image/text_to_image.py \
    --model black-forest-labs/FLUX.1-dev \
    --prompt "a lovely bunny holding a sign that says 'vllm-omni'" \
    --seed 42 --num-inference-steps 50 --guidance-scale 4.0 \
    --height 1024 --width 1024 --enable-cpu-offload \
    --output flux_offload.png

# without offload (same seed)
python examples/offline_inference/text_to_image/text_to_image.py \
    --model black-forest-labs/FLUX.1-dev \
    --prompt "a lovely bunny holding a sign that says 'vllm-omni'" \
    --seed 42 --num-inference-steps 50 --guidance-scale 4.0 \
    --height 1024 --width 1024 \
    --output flux_no_offload.png

# bitwise comparison
python -c "
import numpy as np; from PIL import Image
a = np.array(Image.open('flux_offload.png'))
b = np.array(Image.open('flux_no_offload.png'))
print(f'Bitwise match: {np.array_equal(a, b)}')
print(f'Max pixel diff: {np.abs(a.astype(int)-b.astype(int)).max()}')
print(f'Mean pixel diff: {np.abs(a.astype(int)-b.astype(int)).mean():.4f}')
"
```
</details>

3. **FLUX.1-dev performance** (manual):
   - TP=1 + `--enable-cpu-offload`: before/after timing comparison
   - TP=1/2/4 without offload: regression check

## Test Result

**Hardware**: DGX B200 (NVIDIA B200 192GB)  
**Model**: `black-forest-labs/FLUX.1-dev` (12B params, bf16)  
**Config**: 1024×1024, 50 inference steps, guidance_scale=4.0, seed=42

### CI test

```
tests/e2e/offline_inference/test_diffusion_cpu_offload.py PASSED 
```

### Bitwise correctness (FLUX.1-dev, offload vs no-offload)

```
Bitwise match: True
Max pixel diff: 0
Mean pixel diff: 0.0000
Pixels differ: 0 / 3145728 (0.00%)
```

Output is identical pixel-for-pixel between `--enable-cpu-offload` and no offload.

### Performance checks (TP=1 + CPU offload, 3 reps)

| Config | avg ± std | min | max |
|--------|-----------|-----|-----|
| vllm-omni (before) | 30.65 ± 0.38s | 30.38s | 30.92s |
| vllm-omni (after)  | **24.85 ± 0.39s** | 24.45s | 25.22s |
| diffusers baseline | 25.89 ± 0.35s | 25.64s | 26.13s |

**19% faster** than before, now **4% faster than diffusers**.

### Regression checks (no offload - change has no effect on these paths, just shown as reference)

| Config | Time |
|--------|------|
| TP=1, no offload | 5.65s |
| TP=2, no offload | 4.21s |
| TP=4, no offload | 3.57s |

### Visual comparison

| vllm-omni (offload) | vllm-omni (no offload) | diffusers |
|:---:|:---:|:---:|
| bench_vllm_omni_out | flux_tp1_no_offload | bench_diffusers_out |
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by mengjian0502** [comment]
Very nice!

**Review by hsliuustc0106 — APPROVED** [review]
## Review

**Rating**: 9.5/10 | **Verdict**: ✅ Approved

### Summary
Excellent performance optimization for CPU offloading path, achieving 19% speedup (30.65s → 24.85s) through pre-allocated pinned CPU buffers. Now 4% faster than diffusers baseline.

### Performance Validation (vllm-omni-perf)

| Metric | Before | After | Diffusers | Status |
|--------|--------|-------|-----------|--------|
| CPU offload (TP=1) | 30.65s | **24.85s** | 25.89s | ✅ **19% faster** |
| Bitwise correctness | - | ✅ Match | - | ✅ **Identical output** |
| Regression (no offload) | - | ✅ No impact | - | ✅ **TP=1/2/4 OK** |

### Key Improvements

1. **PinnedWeightsStore**: Pre-allocates page-locked CPU tensors once
2. **Single DMA path**: GPU→CPU copies directly into pinned memory (vs per-parameter pin_memory)
3. **Pre-pin DiT params**: Faster CPU→GPU transfers for modules starting on CPU

### Highlights
- ✅ Comprehensive benchmark data with multiple repetitions
- ✅ Bitwise correctness verification (pixel-perfect match)
- ✅ No regression on non-offload paths (TP=1/2/4)
- ✅ Minimal code change (60 lines, focused optimization)
- ✅ Clear explanation of before/after behavior

### Minor Suggestions (non-blocking)

1. **Memory overhead**: Consider documenting how much extra CPU RAM is consumed by PinnedWeightsStore per module (e.g., "adds ~X GB for FLUX.1-dev").

2. **Fallback path**: The `if store is not None` fallback in `_to_cpu` is good for backward compatibility. Consider adding a debug log when falling back to old path.

3. **Documentation**: PR description mentions "Ref #1686" but could benefit from a brief summary of that issue's root cause for future reference.

### Pitfalls Check

| Directory | Pitfall | Status |
|-----------|---------|--------|
| `diffusion/offloader/` | Memory management | ✅ Verified (pinned stores) |
| `diffusion/offloader/` | Device placement | ✅ Verified (CPU↔GPU) |

### Recommendation
Ready to merge. Outstanding optimization with comprehensive validation.

---

Reviewed by OpenClaw with [vllm-omni-skills](https://github.com/hsliuustc0106/vllm-omni-skills) 🦐

**Skill**: vllm-omni-perf (CPU Offloading)

**Comment by hsliuustc0106** [comment]
this PR is really helpful:)

File: vllm_omni/diffusion/offloader/sequential_backend.py [hunk 1]
```diff
@@ -42,18 +69,21 @@ def _to_cpu(self, module: nn.Module) -> None:
...
+            if self.pin_memory:
+                for p in module.parameters():
+                    if p.data.device.type == "cpu" and not p.data.is_pinned():
+                        p.data = p.data.pin_memory()
 
-        if self.pin_memory:
-            for p in module.parameters():
-                if p.data.device.type == "cpu" and not p.data.is_pinned():
-                    p.data = p.data.pin_memory()
+        torch.cuda.empty_cache()
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/offloader/sequential_backend.py]: does it work in other platform?

File: vllm_omni/diffusion/offloader/sequential_backend.py [hunk 2]
```diff
@@ -42,18 +69,21 @@ def _to_cpu(self, module: nn.Module) -> None:
...
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/offloader/sequential_backend.py:114]: `torch.cuda.empty_cache()` is called unconditionally here. On NPU backends (Ascend), `torch.cuda` doesn't exist and this will fail. The codebase uses `current_omni_platform.is_cuda()` for platform detection (see `vllm_omni/platforms/__init__.py`). Consider:

```python
from vllm_omni.platforms import current_omni_platform

if current_omni_platform.is_cuda():
    torch.cuda.empty_cache()
```

Note: FLUX models are supported on NPU, so this code path will be exercised on non-CUDA backends.

File: vllm_omni/diffusion/offloader/sequential_backend.py [hunk 3]
```diff
@@ -14,6 +16,29 @@
 logger = init_logger(__name__)
 
 
+class PinnedWeightsStore:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/offloader/sequential_backend.py:19]: The `PinnedWeightsStore` allocates pinned CPU buffers but has no cleanup method. If CPU offloading is disabled or the pipeline is reconfigured, these pinned buffers remain allocated. Consider adding a `release()` method to free pinned memory, or ensure the cleanup logic at the end of `enable()` (line 249: "Track modules for cleanup") also releases these stores.

File: vllm_omni/diffusion/offloader/sequential_backend.py [hunk 4]
```diff
@@ -14,6 +16,29 @@
...
+    Allocates page-locked CPU tensors once so that GPU-to-CPU offloading can
+    copy directly into pinned memory (single DMA) instead of the default
+    module.to("cpu") path which creates pageable tensors followed by an
+    expensive per-parameter pin_memory() call.
+    """
+
+    def __init__(self, module: nn.Module):
+        self._store: dict[str, torch.Tensor] = {}
+        for name, tensor in chain(module.named_parameters(), module.named_buffers()):
+            self._store[name] = torch.empty(tensor.shape, dtype=tensor.dtype, device="cpu", pin_memory=True)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/offloader/sequential_backend.py:35]: `pin_memory=True` is CUDA-specific behavior. On NPU, use `torch_npu.npu.mem_pin()` or equivalent. Consider abstracting this via the platform layer if NPU support for CPU offloading is planned.

**Review by hsliuustc0106 — COMMENTED** [review]
## Review Summary

Thanks for the detailed PR with benchmarks and correctness verification! The optimization approach is solid. Found a few hardware compatibility concerns that should be addressed.

### ✅ What's Good
- Performance improvement well-documented (19% faster)
- Bitwise correctness verified
- Existing CI test passes

### ⚠️ Items to Address

| Issue | Severity | Impact |
|-------|----------|--------|
| `torch.cuda.empty_cache()` unconditional | **High** | Breaks on NPU |
| No cleanup for `PinnedWeightsStore` | Medium | Memory leak risk |
| `pin_memory=True` hardcoded | Low | May need NPU equivalent |

Note: FLUX models are supported on NPU (see supported models list), so this code path will be exercised on non-CUDA backends.

File: vllm_omni/diffusion/offloader/sequential_backend.py (hunk 1 above)
Comment (ZhanqiuHu) [reply vllm_omni/diffusion/offloader/sequential_backend.py]: Fixed. Replaced with the guarded version:

```python
empty_cache = current_omni_platform.empty_cache
if empty_cache is not None:
    empty_cache()
```

Consistent with guard pattern from vLLM core ([`parallel_state.py#L1918-L1920`](https://github.com/vllm-project/vllm/blob/fd4a90f33/vllm/distributed/parallel_state.py#L1918-L1920), [`forward_context.py#L397-L399`](https://github.com/vllm-project/vllm/blob/fd4a90f33/vllm/forward_context.py#L397-L399))

This safely proxies to `torch.<device_type>.empty_cache()` on any platform via `Platform.__getattr__`, and is a no-op if the platform doesn't support it. (Note: the original `torch.cuda.empty_cache()` was pre-existing on `main`, not introduced by this PR.)

File: vllm_omni/diffusion/offloader/sequential_backend.py (hunk 3 above)
Comment (ZhanqiuHu) [reply vllm_omni/diffusion/offloader/sequential_backend.py:19]: Added `PinnedWeightsStore.release()` method that clears the buffer dict. Stores are now tracked as `self._pinned_stores` on `ModelLevelOffloadBackend` and released in `disable()`.

File: vllm_omni/diffusion/offloader/sequential_backend.py (hunk 4 above)
Comment (ZhanqiuHu) [reply vllm_omni/diffusion/offloader/sequential_backend.py:35]: Good point. In updated code, `PinnedWeightsStore` creation only happens if `current_omni_platform.is_cuda_alike()` now, so NPU/XPU never hit this path, and they just fall back to plain `module.to("cpu")` with a warning.

As for abstracting via the platform layer, I think we could do that as a follow-up once there's an actual NPU consumer to test against, so we don't end up designing the wrong interface. When that happens, we could turn `PinnedWeightsStore` into an ABC with platform-specific subclasses and rename `offload_from_gpu` ->`offload_from_device`. I added a docstring note flagging this as the extension point.

**Comment by ZhanqiuHu** [comment]
Rebased and resolved conflicts from upstream. Ready for merging now.

File: vllm_omni/diffusion/offloader/sequential_backend.py [hunk 5]
```diff
@@ -63,13 +98,19 @@ def _to_cpu(self, module: nn.Module) -> None:
...
+                for p in module.parameters():
+                    if p.data.device.type == "cpu" and not p.data.is_pinned():
+                        p.data = p.data.pin_memory()
 
-        if self.pin_memory:
-            for p in module.parameters():
-                if p.data.device.type == "cpu" and not p.data.is_pinned():
-                    p.data = p.data.pin_memory()
+        empty_cache = current_omni_platform.empty_cache
+        if empty_cache is not None:
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/offloader/sequential_backend.py:112]: Prefer to use the following directly
```python
current_omni_platform.empty_cache()
```

File: vllm_omni/diffusion/offloader/sequential_backend.py [hunk 6]
```diff
@@ -202,12 +250,38 @@ def enable(self, pipeline: nn.Module) -> None:
...
+                        p.data = p.data.pin_memory()
+
+            # Pre-allocate pinned CPU buffers for encoders (currently on GPU)
+            # so GPU->CPU offload copies directly into pinned memory
+            for enc in modules.encoders:
+                self._pinned_stores[id(enc)] = PinnedWeightsStore(enc)
+                logger.debug(
+                    "Pre-allocated pinned CPU buffers for %s",
+                    enc.__class__.__name__,
+                )
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/offloader/sequential_backend.py:276]: Shall we move the creation and assignment of PinnedWeightsStore instances from `ModelLevelOffloadBackend` to `SequentialOffloadHook`? Thus we don't need to pass stores through multiple layers.

File: vllm_omni/diffusion/offloader/sequential_backend.py [hunk 7]
```diff
@@ -202,12 +250,38 @@ def enable(self, pipeline: nn.Module) -> None:
...
+            # Pre-pin DiT parameters that start on CPU for fast CPU->GPU transfer
+            for dit_mod in modules.dits:
+                for p in dit_mod.parameters():
+                    if p.data.device.type == "cpu" and not p.data.is_pinned():
+                        p.data = p.data.pin_memory()
+
+            # Pre-allocate pinned CPU buffers for encoders (currently on GPU)
+            # so GPU->CPU offload copies directly into pinned memory
+            for enc in modules.encoders:
+                self._pinned_stores[id(enc)] = PinnedWeightsStore(enc)
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/offloader/sequential_backend.py:272]: It seems only encoder modules' PinnedWeightsStore are initialized and stored into `_pinned_stores`, what about those for DiT modules? Are we going to apply the same pre-allocation method for dit modules

Comment (ZhanqiuHu) [reply vllm_omni/diffusion/offloader/sequential_backend.py:272]: Thanks for the review feedback. Just re-benchmarked and found that pre-pinning no longer provides perf gain. #1723 introduced `_move_params` which replaced the slow recursive `module.to("cpu")` with direct per-parameter `.to(device, non_blocking=True)`. That was the same core bottleneck `PinnedWeightsStore` was working around. I will close this PR.

**Comment by ZhanqiuHu** [comment]
Closing this since the improvement is subsumed by https://github.com/vllm-project/vllm-omni/pull/1723

**Comment by nuclearwu** [comment]
@ZhanqiuHu I tried running the Flux.2-dev model with this PR and https://github.com/vllm-project/vllm-omni/pull/1723. Your version only takes 82.1101 seconds, while https://github.com/vllm-project/vllm-omni/pull/1723 requires 94.9571 seconds. So, should we consider it?

**Comment by yuanheng-zhao** [comment]
The scope of #1723 is to support cpu offloading as well as cache-dit to be enabled together, but I don't think it covers the handling or improvement of `pin_memory`. Your PR still improves the performance by deduplicate the calling of `pin_memory`.

Would you like to rebase and test again? Thanks for contributing!

**Comment by ZhanqiuHu** [comment]
Thanks @nuclearwu for testing on Flux.2-dev and @yuanheng-zhao for the feedback.

I re-benchmarked on H100 with FLUX.1-dev (1024x1024, 50 steps, guidance_scale=4.0, 6 gens, first discarded as warmup) and saw no steady-state difference, which is why I closed:

| | main (post-#1723) | this PR |
|---|---|---|
| enable() init | 3s | 24s |
| gen 1 (warmup) | 20.65s | 9.15s |
| gen 2-6 avg | 9.4s | 9.5s |

On FLUX.1-dev the pre-pinning cost just shifts from gen 1 into `enable()`, no net win. 

So @nuclearwu, can you try adding a warmup run before your actual run to see if the perf diff disappears? I suspect the 82.1s vs 94.9s difference might also be the cold-start effect rather than a real steady-state improvement.


Side note: I also tried adding PinnedWeightsStore for DiT modules. RSS went from ~50GB to ~82GB (+32GB) with zero speed improvement, not even on the warmup run.
